crates/cadenza/src/mcp.rs

```diff
@@ -0,0 +1,430 @@
+                Value::Float(_) => "Float".to_string(),
+                Value::String(_) => "String".to_string(),
+                Value::List(_) => "List".to_string(),
+                Value::Record { .. } => "Record".to_string(),
+                Value::StructConstructor { name, .. } => format!("StructConstructor<{}>", name),
+                Value::Type(_) => "Type".to_string(),
+                Value::Quantity { dimension, .. } => format!("Quantity<{}>", dimension),
+                Value::UnitConstructor(_) => "UnitConstructor".to_string(),
+                Value::BuiltinFn(_) => "BuiltinFn".to_string(),
+                Value::BuiltinMacro(_) => "BuiltinMacro".to_string(),
+                Value::SpecialForm(_) => "SpecialForm".to_string(),
+                Value::UserFunction(_) => "UserFunction".to_string(),
+            }
+        } else {
+            "Nil".to_string()
+        };
+        Ok(CallToolResult::success(vec![Content::text(format!(
+            "Type: {}",
+            type_str
+        ))]))
+    }
+
+    /// Check dimensional analysis for an expression with units
+    #[tool(description = "Check dimensional analysis for an expression with units. Use this to verify unit compatibility and conversions.")]
+    async fn check_dimensions(
+        &self,
+        rmcp::handler::server::wrapper::Parameters(
+            req,
+        ): rmcp::handler::server::wrapper::Parameters<CheckDimensionsRequest>,
+    ) -> Result<CallToolResult, McpError> {
+        let mut compiler = Compiler::new();
+        let mut env = Env::new();
+
+        let parsed = parse(&req.expression);
+        if !parsed.errors.is_empty() {
+            return Ok(CallToolResult::error(vec![Content::text(
+                "Parse errors encountered".to_string(),
+            )]));
+        }
+
+        let result = cadenza_eval::eval(&parsed.ast(), &mut env, &mut compiler);
+        // Check the last value
+        if let Some(value) = result.last() {
+            match value {
+                Value::Quantity {
+                    value: val,
+                    unit,
+                    dimension,
+                } => Ok(CallToolResult::success(vec![Content::text(format!(
+                    "Value: {}\nDimension: {}\nUnit: {:?}",
+                    val, dimension, unit
+                ))])),
+                _ => Ok(CallToolResult::success(vec![Content::text(
+                    "Expression has no dimensions (dimensionless value)".to_string(),
+                )])),
+            }
+        } else {
+            Ok(CallToolResult::success(vec![Content::text(
+                "No value produced".to_string(),
+            )]))
+        }
+    }
+
+    /// List all built-in functions and operators
+    #[tool(description = "List all built-in functions, operators, and special forms available in Cadenza. Use this to discover available functionality.")]
+    async fn list_builtins(&self) -> Result<CallToolResult, McpError> {
+        let builtins = [
```

Comment:
This leaves a lot to be desired. It should really query the std environment to get this information. We also need to have docs on of these so it's clear how to use them.